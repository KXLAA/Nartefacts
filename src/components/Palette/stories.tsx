import { Story, Meta } from '@storybook/react'

import { Palette } from '.'

export default {
  component: Palette,
  title: 'Palette',
} as Meta

export const Basic: Story = () => <Palette colors={[]} />
