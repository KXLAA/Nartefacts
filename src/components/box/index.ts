export { Box } from './Box'
export type { BoxProps } from './box.types'
