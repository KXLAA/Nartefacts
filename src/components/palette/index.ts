export { Palette } from './Palette'
export type { ColorsTuple } from './types'
