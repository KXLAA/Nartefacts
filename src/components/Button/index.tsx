import * as S from './styles'
import { capitalize } from 'utils/utils'

export type ButtonProps = {
  text?: string
  buttonColor?: string
  textColor?: string
}

export const Button: React.FC<ButtonProps> = ({
  buttonColor,
  text,
  textColor,
}) => {
  return (
    <S.Button buttonColor={buttonColor} textColor={textColor} role="button">
      {capitalize(text!) || 'Button'}
    </S.Button>
  )
}
