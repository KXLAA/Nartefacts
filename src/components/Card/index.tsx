import Image from 'next/image'

import * as S from '@/components/Card/styles'

export type CardProps = {
  image?: string | null
  colors?: string[] | null
  likeCount?: number | null
  textColor?: string | null
  albumArt?: string | null
  spotify?: string | null
  apple?: string | null
  type?: string | null
  description?: string | null
  title?: string | null
  artist?: {
    name?: string | null
    photoUrl?: string | null
    biography?: string | null
  } | null
}

export const Card: React.FC<CardProps> = ({ albumArt }) => {
  return (
    <S.Card>
      <S.Gradient />
      <S.ImageWrapper>
        <Image
          src={albumArt!}
          height={385}
          width={385}
          alt={'album art'}
          layout="responsive"
        />
      </S.ImageWrapper>
    </S.Card>
  )
}
