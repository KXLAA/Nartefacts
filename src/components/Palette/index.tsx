import toast from 'react-hot-toast'

import * as Layout from '@/components/Common/Layout'
import * as S from '@/components/Palette/styles'
import { useCopyToClipboard, useMouseOver } from '@/lib/hooks'

export type PalletteProps = {
  colors: string[]
  small?: boolean
}

export type ColorBoxProps = {
  color: string
  small?: boolean
}

export type ColorsTuple = [
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
]

export const ColorBox: React.FC<ColorBoxProps> = ({ color, small }) => {
  const [, copy] = useCopyToClipboard()
  const [hoverRef, isHovered] = useMouseOver()

  const copyToClipboard = (color: string) => {
    copy(color)
    toast(`Copied ${color}`)
  }

  return (
    <S.Color
      small={small}
      ref={hoverRef as any}
      css={{
        backgroundColor: color.match(
          /^#(?:(?:[\da-f]{3}){1,2}|(?:[\da-f]{4}){1,2})$/i,
        )
          ? color
          : '#202020',
      }}
      onClick={() => copyToClipboard(color)}
      data-testid="color-box"
    >
      <span>{isHovered && `COPY`}</span>
    </S.Color>
  )
}

export const Palette: React.FC<PalletteProps> = ({ colors, small }) => {
  colors.forEach((color, index) => {
    if (!color.match(/^#(?:(?:[\da-f]{3}){1,2}|(?:[\da-f]{4}){1,2})$/i)) {
      console.warn(
        `Invalid color ${color} at index ${index} in Palette component.`,
      )
    }
  })

  return (
    <Layout.Grid columns={4} gap={small ? 1 : 2}>
      {colors?.map((color) => (
        <ColorBox key={color} color={color} small={small} />
      ))}
    </Layout.Grid>
  )
}
