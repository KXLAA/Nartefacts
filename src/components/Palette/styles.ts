import styled from 'styled-components'
import { getColor, getFontSize } from '@/styles/global'
import { ColorBoxProps } from '@/components/Palette'

export const Color = styled.div<ColorBoxProps>`
  height: ${({ small }) => (small === true ? '58px' : '130px')};
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  border-radius: ${({ small }) => (small ? '4px' : '8px')};
  transition: all 0.3s ease-in-out;
  background: ${({ color }) =>
    color.match(/^#(?:(?:[\da-f]{3}){1,2}|(?:[\da-f]{4}){1,2})$/i)
      ? color
      : getColor('blackLight')};
  :hover {
    filter: brightness(150%);
  }
  span {
    font-weight: bold;
    font-size: ${({ small }) =>
      small ? getFontSize('sm') : getFontSize('lg')};
    color: ${getColor('white')};
  }
`
