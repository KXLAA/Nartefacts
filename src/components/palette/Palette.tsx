import toast from 'react-hot-toast'

import { useAutoAnimate, useCopyToClipboard, useMouseOver } from '@/lib/hooks'

import { StyledColor, StyledPalette } from './styles'
import { ColorBoxProps, PalletteProps } from './types'

export const ColorBox = ({ color, small }: ColorBoxProps) => {
  const [, copy] = useCopyToClipboard()
  const [hoverRef, isHovered] = useMouseOver()

  const copyToClipboard = (color: string) => {
    copy(color)
    toast(`Copied ${color}`)
  }

  return (
    <StyledColor
      small={small}
      ref={hoverRef as any}
      css={{
        backgroundColor: color.match(
          /^#(?:(?:[\da-f]{3}){1,2}|(?:[\da-f]{4}){1,2})$/i,
        )
          ? color
          : '#202020',
      }}
      onClick={() => copyToClipboard(color)}
      data-testid="color-box"
    >
      <span>{isHovered && `COPY`}</span>
    </StyledColor>
  )
}

export const Palette = ({ colors, small }: PalletteProps) => {
  const [parent] = useAutoAnimate()
  colors?.forEach((color, index) => {
    if (!color.match(/^#(?:(?:[\da-f]{3}){1,2}|(?:[\da-f]{4}){1,2})$/i)) {
      console.warn(
        `Invalid color ${color} at index ${index} in Palette component.`,
      )
    }
  })

  return (
    <StyledPalette
      small={small}
      ref={parent as React.RefObject<HTMLDivElement>}
    >
      {colors?.slice(small ? 4 : undefined).map((color) => (
        <ColorBox key={color} color={color} small={small} />
      ))}
    </StyledPalette>
  )
}
