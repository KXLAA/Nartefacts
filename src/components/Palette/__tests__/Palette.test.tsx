import { render } from '@testing-library/react'
import { colorsTuple } from 'components/Palette'
import { colors as testColors } from 'components/utils'

import { Palette } from '..'

describe('<Palette />', () => {
  it('should render the pallette', () => {
    const { getByTitle } = render(<Palette colors={testColors} />)
    const colors = getByTitle(/palette/i)
    expect(colors).toBeInTheDocument()
  })

  it('should render the pallette correctly', () => {
    const { container } = render(<Palette colors={testColors} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the pallette with correct color props', () => {
    const { getAllByTitle } = render(<Palette colors={testColors} />)
    const colors = getAllByTitle(/color/i)
    expect(colors).toHaveLength(testColors.length)

    colors.forEach((color, index) => {
      expect(color).toHaveStyle(`background: ${testColors[index]}`)
    })
  })

  it('should render the default color if invalid color is passed', () => {
    const invalidColors: colorsTuple = Object.assign([], testColors, {
      0: '0955',
    })
    const { getAllByTitle } = render(<Palette colors={invalidColors} />)
    const colors = getAllByTitle(/color/i)
    expect(colors).toHaveLength(testColors.length)
    expect(colors[0]).toHaveStyle(`background: #202020`)
  })
})
