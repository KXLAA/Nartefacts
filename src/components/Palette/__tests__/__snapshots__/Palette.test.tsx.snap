// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Palette /> should render the pallette correctly 1`] = `
.c0 {
  display: grid;
  grid-template-columns: repeat(4,1fr);
  gap: 0.5rem;
  width: 100%;
  max-width: 800px;
}

.c1 {
  height: 130px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  border-radius: 8px;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  background: #095525;
}

.c1:hover {
  -webkit-filter: brightness(150%);
  filter: brightness(150%);
}

.c2 {
  height: 130px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  border-radius: 8px;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  background: #673636;
}

.c2:hover {
  -webkit-filter: brightness(150%);
  filter: brightness(150%);
}

.c3 {
  height: 130px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  border-radius: 8px;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  background: #379273;
}

.c3:hover {
  -webkit-filter: brightness(150%);
  filter: brightness(150%);
}

.c4 {
  height: 130px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  border-radius: 8px;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  background: #136407;
}

.c4:hover {
  -webkit-filter: brightness(150%);
  filter: brightness(150%);
}

.c5 {
  height: 130px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  border-radius: 8px;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  background: #573227;
}

.c5:hover {
  -webkit-filter: brightness(150%);
  filter: brightness(150%);
}

.c6 {
  height: 130px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  border-radius: 8px;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  background: #547820;
}

.c6:hover {
  -webkit-filter: brightness(150%);
  filter: brightness(150%);
}

.c7 {
  height: 130px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  border-radius: 8px;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  background: #163598;
}

.c7:hover {
  -webkit-filter: brightness(150%);
  filter: brightness(150%);
}

.c8 {
  height: 130px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  border-radius: 8px;
  -webkit-transition: all 0.3s ease-in-out;
  transition: all 0.3s ease-in-out;
  background: #682052;
}

.c8:hover {
  -webkit-filter: brightness(150%);
  filter: brightness(150%);
}

<div
  class="c0"
  title="Palette"
>
  <div
    class="c1"
    color="#095525"
    title="Color"
  />
  <div
    class="c2"
    color="#673636"
    title="Color"
  />
  <div
    class="c3"
    color="#379273"
    title="Color"
  />
  <div
    class="c4"
    color="#136407"
    title="Color"
  />
  <div
    class="c5"
    color="#573227"
    title="Color"
  />
  <div
    class="c6"
    color="#547820"
    title="Color"
  />
  <div
    class="c7"
    color="#163598"
    title="Color"
  />
  <div
    class="c8"
    color="#682052"
    title="Color"
  />
</div>
`;
